### PR TITLE
[PfemFluid] 2d and 3d inlet 

### PR DESCRIPTION
**Pull request by AFranci**
This PR improves the current version of inlet boundaries model used in PFEM. It also adds the 3D version. Attention is paid to avoid artificial topological changes in the pfem domain and to maintain the original mesh sizes in the different region of the domain.

### CODE REVIEW

File: applications/PfemFluidDynamicsApplication/custom_processes/build_mesh_boundary_for_fluids_process.hpp [hunk 1]
```diff
@@ -104,7 +104,7 @@ namespace Kratos
 
       bool success = false;
 
-      //boost::timer auxiliary;
+      // boost::timer auxiliary;
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/build_mesh_boundary_for_fluids_process.hpp]: maybe remove?

File: applications/PfemFluidDynamicsApplication/custom_processes/build_mesh_boundary_for_fluids_process.hpp [hunk 2]
```diff
@@ -373,7 +373,7 @@ namespace Kratos
     {
       KRATOS_TRY
 
-      //rModelPart.AddCondition(pCondition); //if a Local Id corresponds to a Global Id not added
+      // rModelPart.AddCondition(pCondition); //if a Local Id corresponds to a Global Id not added
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/build_mesh_boundary_for_fluids_process.hpp]: remove?

File: applications/PfemFluidDynamicsApplication/custom_processes/build_mesh_boundary_for_fluids_process.hpp [hunk 3]
```diff
@@ -489,7 +489,7 @@ namespace Kratos
     BuildMeshBoundaryForFluidsProcess &operator=(BuildMeshBoundaryForFluidsProcess const &rOther);
 
     /// Copy constructor.
-    //BuildMeshBoundaryForFluidsProcess(BuildMeshBoundaryForFluidsProcess const& rOther);
+    // BuildMeshBoundaryForFluidsProcess(BuildMeshBoundaryForFluidsProcess const& rOther);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/build_mesh_boundary_for_fluids_process.hpp]: remove

File: applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp [hunk 4]
```diff
@@ -1648,6 +1211,40 @@ namespace Kratos
 						}
 					}
 				}
+
+				if (eulerianInletNodes > 0 && LargestEdge > (1.7 * meanMeshSize))
+				{
+
+					array_1d<double, 3> NewPosition = (Element[FirstEdgeNode[maxCount]].Coordinates() + Element[SecondEdgeNode[maxCount]].Coordinates()) * 0.5;
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp]: `new_position`

File: applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp [hunk 5]
```diff
@@ -900,10 +918,13 @@ namespace Kratos
 				if ((fabs(cosAngle12) > 0.995 || fabs(cosAngle13) > 0.995 || fabs(cosAngle14) > 0.995) && (cosAngleBetweenNormals01 > 0.99 && cosAngleBetweenNormals02 > 0.99 && cosAngleBetweenNormals12 > 0.99))
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp]: `std::abs` better

File: applications/PfemFluidDynamicsApplication/pfem_fluid_dynamics_application_variables.cpp [hunk 6]
```diff
@@ -98,6 +91,11 @@ KRATOS_CREATE_VARIABLE(double, SOLID_VOLUMETRIC_COEFFICIENT)
 KRATOS_CREATE_VARIABLE(double, SOLID_DEVIATORIC_COEFFICIENT)
 KRATOS_CREATE_VARIABLE(bool, SOLID_INTERFACE_NODE)
 
+KRATOS_CREATE_LOCAL_FLAG(PFEMFlags, EULERIAN_INLET, 0);
+KRATOS_CREATE_LOCAL_FLAG(PFEMFlags, LAGRANGIAN_INLET, 1);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/pfem_fluid_dynamics_application_variables.cpp:91]: better solution indeed

File: applications/PfemFluidDynamicsApplication/python_scripts/pfem_fluid_dynamics_analysis.py [hunk 7]
```diff
@@ -398,17 +398,13 @@ def AddPfemVariables(self):
         self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.REACTION)
         self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.NORMAL)
         self.main_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.VOLUME_ACCELERATION)
-        self.main_model_part.AddNodalSolutionStepVariable(KratosPfemFluid.FREESURFACE)
-        self.main_model_part.AddNodalSolutionStepVariable(KratosPfemFluid.PREVIOUS_FREESURFACE)
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/python_scripts/pfem_fluid_dynamics_analysis.py:402]: maybe I should remove those two in the `Fem2Dem`?

**Review by AlejandroCornejo — CHANGES_REQUESTED** [review]

File: applications/PfemFluidDynamicsApplication/python_scripts/pfem_fluid_dynamics_analysis.py (hunk 7 above)
Comment (AFranci) [reply applications/PfemFluidDynamicsApplication/python_scripts/pfem_fluid_dynamics_analysis.py:402]: yes!

File: applications/PfemFluidDynamicsApplication/custom_processes/build_mesh_boundary_for_fluids_process.hpp [hunk 8]
```diff
@@ -195,14 +187,14 @@ namespace Kratos
       }
       const unsigned int dimension = mrModelPart.ElementsBegin()->GetGeometry().WorkingSpaceDimension();
 
-      //reset the boundary flag in all nodes and check if a remesh process has been performed
+      // reset the boundary flag in all nodes and check if a remesh process has been performed
       bool any_node_to_erase = false;
       for (ModelPart::NodesContainerType::const_iterator in = rModelPart.NodesBegin(); in != rModelPart.NodesEnd(); in++)
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/build_mesh_boundary_for_fluids_process.hpp:192]: I suppose that this cannot be done in OpenMP?

File: applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp [hunk 9]
```diff
@@ -102,6 +102,7 @@ namespace Kratos
 			const ProcessInfo &rCurrentProcessInfo = mrModelPart.GetProcessInfo();
 			double currentTime = rCurrentProcessInfo[TIME];
 			double timeInterval = rCurrentProcessInfo[DELTA_TIME];
+			const unsigned int dimension = mrModelPart.ElementsBegin()->GetGeometry().WorkingSpaceDimension();
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp]: we usually use `std::size`_t for this purpose

File: applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp [hunk 10]
```diff
@@ -2074,9 +1671,9 @@ namespace Kratos
 					pnode->pAddDof(rDof);
 				}
 
-				Node<3>::Pointer SlaveNode1 = mrModelPart.pGetNode(NodesIDToInterpolate[nn][0]);
-				Node<3>::Pointer SlaveNode2 = mrModelPart.pGetNode(NodesIDToInterpolate[nn][1]);
-				Node<3>::Pointer SlaveNode3 = mrModelPart.pGetNode(NodesIDToInterpolate[nn][2]);
+				Node<3>::Pointer SlaveNode1 = mrModelPart.pGetNode(nodes_id_to_interpolate[nn][0]);
+				Node<3>::Pointer SlaveNode2 = mrModelPart.pGetNode(nodes_id_to_interpolate[nn][1]);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp]: better to use `auto slave_node_2`

File: applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp [hunk 11]
```diff
@@ -2175,8 +1772,8 @@ namespace Kratos
 
 				list_of_new_nodes.push_back(pnode);
 
-				Node<3>::Pointer SlaveNode1 = mrModelPart.pGetNode(NodesIDToInterpolate[nn][0]);
-				Node<3>::Pointer SlaveNode2 = mrModelPart.pGetNode(NodesIDToInterpolate[nn][1]);
+				Node<3>::Pointer SlaveNode1 = mrModelPart.pGetNode(nodes_id_to_interpolate[nn][0]);
+				Node<3>::Pointer SlaveNode2 = mrModelPart.pGetNode(nodes_id_to_interpolate[nn][1]);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/generate_new_nodes_before_meshing_process.hpp]: same with `auto`

File: applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp [hunk 12]
```diff
@@ -408,14 +398,14 @@ namespace Kratos
 				if (in->IsNot(NEW_ENTITY) && in->IsNot(INLET) && in->IsNot(ISOLATED) && in->IsNot(RIGID) && in->IsNot(SOLID))
 				{
 					unsigned int neighErasedNodes = 0;
-					radius = 0.6 * meshSize;
+					unsigned int freeSurfaceNeighNodes = 0;
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp]: std::size_t

File: applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp [hunk 13]
```diff
@@ -533,14 +523,18 @@ namespace Kratos
...
 									{
 										mrRemesh.Info->BalancePrincipalSecondaryPartsNodes += -1;
 									}
 								}
 							}
 						}
-						else if (in->IsNot(INLET))
+						else
 						{
 							unsigned int k = 0;
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp]: `sid::size_t`

File: applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp [hunk 14]
```diff
@@ -809,6 +817,12 @@ namespace Kratos
 				}
 			}
 
+			bool inletElement = false;
+			if (eElement[0].Is(PFEMFlags::EULERIAN_INLET) || eElement[1].Is(PFEMFlags::EULERIAN_INLET) || eElement[2].Is(PFEMFlags::EULERIAN_INLET) || eElement[3].Is(PFEMFlags::EULERIAN_INLET))
+			{
+				inletElement = true;
+			}
+
 			unsigned int rigidNode = 0;
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp]: sid::size_t

File: applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp [hunk 15]
```diff
@@ -933,10 +954,13 @@ namespace Kratos
...
-							// std::cout<<"erased an isolated element node"<<std::endl;
-							erased_nodes += 1;
 							inside_nodes_removed++;
 
+							if (inletElement)
+							{
+								nodes_removed_inlet_zone++;
+							}
+
 							const unsigned int propertyIdNode = eElement[i].FastGetSolutionStepValue(PROPERTY_ID);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp]: `sid::size_t`

File: applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp [hunk 16]
```diff
@@ -961,10 +985,13 @@ namespace Kratos
...
-										// std::cout<<"_________________________          erased an isolated element node"<<std::endl;
-										erased_nodes += 1;
 										inside_nodes_removed++;
 
+										if (inletElement)
+										{
+											nodes_removed_inlet_zone++;
+										}
+
 										const unsigned int propertyIdNode = eElement[i].FastGetSolutionStepValue(PROPERTY_ID);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/remove_mesh_nodes_for_fluids_process.hpp]: `const sid::size_t`

File: applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp [hunk 17]
```diff
@@ -162,7 +162,7 @@ namespace Kratos
                     unsigned int numfreesurf = 0;
                     unsigned int numboundary = 0;
                     unsigned int numrigid = 0;
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp]: `sid::size_t`

File: applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp [hunk 18]
```diff
@@ -583,7 +588,7 @@ namespace Kratos
                     //             CriticalVolume = 0.00001 * regularTetrahedronVolume;
                     //         }
 
-                    //         if (fabs(Volume) < CriticalVolume)
+                    //         if (std::abs(Volume) < CriticalVolume)
                     //         {
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_processes/select_mesh_elements_for_fluids_process.hpp]: remove commented code :)

File: applications/PfemFluidDynamicsApplication/custom_strategies/strategies/v_p_strategy.h [hunk 19]
```diff
@@ -402,8 +402,8 @@ namespace Kratos
       for (ModelPart::NodeIterator i = rModelPart.NodesBegin();
            i != rModelPart.NodesEnd(); ++i)
       {
+        if ((i)->IsNot(PFEMFlags::EULERIAN_INLET))
 
-        if ((i)->GetValue(EULERIAN_INLET) == false)
         {
           array_1d<double, 3> &CurrentVelocity = (i)->FastGetSolutionStepValue(VELOCITY, 0);
```
Comment (AlejandroCornejo) [inline applications/PfemFluidDynamicsApplication/custom_strategies/strategies/v_p_strategy.h:408]: `auto& r_current_velocity`

**Review by AlejandroCornejo — APPROVED** [review]

**Pull request merged**